[quant][core][feature] Implement masked_fill for CUDA tensors

jcaip opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #85164
* __->__ #85108

Summary:
- Add new cuda test for masked_fill
- Add in QuantizedCUDA implementation for masked_fill

Test Plan:
```
python test/test_quantization.py -k test_qtensor_masked_fill_cuda

```

Reviewers: dzdang

Subscribers:

Tasks: Fixes https://github.com/pytorch/pytorch/issues/83110

Tags: quant

jcaip committed:
[quant][core][feature] Implement masked_fill for CUDA tensors

Summary:
- Add new cuda test for masked_fill
- Add in QuantizedCUDA implementation for masked_fill

Test Plan:
```
python test/test_quantization.py -k test_qtensor_masked_fill_cuda

```

Reviewers: @dzdang

Subscribers:

Tasks: Fixes https://github.com/pytorch/pytorch/issues/83110

Tags: quant

[ghstack-poisoned]

vadimkantorov commented on test/quantization/core/test_quantized_tensor.py:
```diff
@@ -1032,8 +1037,7 @@ def test_qtensor_masked_fill(self):
             # mask fill the whole tensor, equivalent to calling plain vanilla fill
-            mask = torch.tensor(True)
-            torch.tensor(fill_with)
+            mask = torch.tensor(True, device=device)
```
does masked_fill_ also work with python scalar? e.g. q_masked_fill.masked_fill_(True, fill_with)

jcaip replied:
I don't think so, `masked_fill` expects a tensor mask, so it will throw a type error.

vadimkantorov replied:
my bad, i got it mistaken for the fill value

dzdang commented on test/quantization/core/test_quantized_tensor.py:
```diff
@@ -1049,18 +1053,19 @@ def test_qtensor_masked_fill(self):
                 if mask[i]:
                     # this assignment doesn't end up calling masked_fill, allowing us to compare the different implementations
-                    ref[i] = fill_with
+                    ref[i] = torch.tensor([fill_with], device=device, dtype=torch.float)
```
btw you can also just `do torch.tensor(fill_with, device=device, dtype=torch.float)`

dzdang reviewed (approved):
lgtm

jcaip commented:
@pytorchbot merge

jcaip committed:
Update on "[quant][core][feature] Implement masked_fill for quantized CUDA tensors"


Summary:
- Add new cuda test for masked_fill
- Add in QuantizedCUDA implementation for masked_fill

Test Plan:
```
python test/test_quantization.py -k test_qtensor_masked_fill_cuda

```

Reviewers: dzdang

Subscribers:

Tasks: 

Tags: quant

[ghstack-poisoned]

vadimkantorov commented on test/quantization/core/test_quantized_tensor.py:
```diff
@@ -1049,18 +1053,19 @@ def test_qtensor_masked_fill(self):
         # the above loop does the same test as test_qtensor_fill
         # now we will check masked_fill for subset of indices
-        mask = torch.randint(0, 2, (numel, ))
+        mask = torch.randint(0, 2, (numel, ), device=device)
```
related : https://github.com/pytorch/pytorch/issues/35527

jcaip committed:
Update on "[quant][core][feature] Implement masked_fill for quantized CUDA tensors"


Summary:
- Add new cuda test for masked_fill
- Add in QuantizedCUDA implementation for masked_fill

Test Plan:
```
python test/test_quantization.py -k test_qtensor_masked_fill_cuda

```

Reviewers: dzdang

Subscribers:

Tasks: 

Tags: quant

[ghstack-poisoned]

jcaip committed:
Update on "[quant][core][feature] Implement masked_fill for CUDA tensors"


Summary:
- Add new cuda test for masked_fill
- Add in QuantizedCUDA implementation for masked_fill

Test Plan:
```
python test/test_quantization.py -k test_qtensor_masked_fill_cuda

```

Reviewers: dzdang

Subscribers:

Tasks: Fixes https://github.com/pytorch/pytorch/issues/83110

Tags: quant

[ghstack-poisoned]

jcaip commented:
@pytorchbot merge
